<a href="https://colab.research.google.com/github/markPranav/Machine-Learning/blob/main/160119733113_Lab_Internal_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 160119733113 - Machine Learning Lab Internal 1

## Name - Sripranav Muktevi




## AIM: Polynomial Regression on Air Quality to predict Absolute Humidity
**Dataset** : https://archive-beta.ics.uci.edu/ml/datasets/air+quality
### Description

Polynomial Regression is a regression algorithm that models the relationship between a dependent(y) and independent variable(x) as nth degree polynomial. The Polynomial Regression equation is given below:

`y= b0+b1x1+ b2x12+ b2x13+...... bnx1n`

It is also called the special case of Multiple Linear Regression in ML. Because we add some polynomial terms to the Multiple Linear regression equation to convert it into Polynomial Regression.

It is a linear model with some modification in order to increase the accuracy.
The dataset used in Polynomial regression for training is of non-linear nature.
It makes use of a linear regression model to fit the complicated and non-linear functions and datasets.

Hence, "In Polynomial regression, the original features are converted into Polynomial features of required degree (2,3,..,n) and then modeled using a linear model."
### Attribute Information:

- Date (DD/MM/YYYY) 
- Time (HH.MM.SS) 
-  True hourly averaged concentration CO in mg/m^3 (reference analyzer) 
-  PT08.S1 (tin oxide) hourly averaged sensor response (nominally CO targeted) 
-  True hourly averaged overall Non Metanic HydroCarbons concentration in microg/- ^3 (reference analyzer) 
-  True hourly averaged Benzene concentration in microg/m^3 (reference analyzer) 
-  PT08.S2 (titania) hourly averaged sensor response (nominally NMHC targeted) 
-  True hourly averaged NOx concentration in ppb (reference analyzer) 
-  PT08.S3 (tungsten oxide) hourly averaged sensor response (nominally NOx targeted) 
- True hourly averaged NO2 concentration in microg/m^3 (reference analyzer) 
-  PT08.S4 (tungsten oxide) hourly averaged sensor response (nominally NO2 targeted) 
-  PT08.S5 (indium oxide) hourly averaged sensor response (nominally O3 - rgeted) 
-  Temperature in °C 
-  Relative Humidity (%) 
-  AH Absolute Humidity

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn
from matplotlib import style
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures 
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import operator


In [2]:
data_set= pd.read_csv('AirQualityUCI.csv', sep=";" , parse_dates= ['Date','Time'])
data_set.describe()

PT08.S1(CO)     NMHC(GT)  PT08.S2(NMHC)      NOx(GT)  PT08.S3(NOx)  \
count  9357.000000  9357.000000    9357.000000  9357.000000   9357.000000   
mean   1048.990061  -159.090093     894.595276   168.616971    794.990168   
std     329.832710   139.789093     342.333252   257.433866    321.993552   
min    -200.000000  -200.000000    -200.000000  -200.000000   -200.000000   
25%     921.000000  -200.000000     711.000000    50.000000    637.000000   
50%    1053.000000  -200.000000     895.000000   141.000000    794.000000   
75%    1221.000000  -200.000000    1105.000000   284.000000    960.000000   
max    2040.000000  1189.000000    2214.000000  1479.000000   2683.000000   

           NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  Unnamed: 15  Unnamed: 16  
count  9357.000000   9357.000000  9357.000000          0.0          0.0  
mean     58.148873   1391.479641   975.072032          NaN          NaN  
std     126.940455    467.210125   456.938184          NaN          NaN  
min    -200.000000   -200.000000  -200.000000          NaN          NaN  
25%      53.000000   1185.000000   700.000000          NaN          NaN  
50%      96.000000   1446.000000   942.000000          NaN          NaN  
75%     133.000000   1662.000000  1255.000000          NaN          NaN  
max     340.000000   2775.000000  2523.000000          NaN          NaN

In [3]:
data_set.head()

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0 2004-10-03  18.00.00    2,6       1360.0     150.0     11,9         1046.0   
1 2004-10-03  19.00.00      2       1292.0     112.0      9,4          955.0   
2 2004-10-03  20.00.00    2,2       1402.0      88.0      9,0          939.0   
3 2004-10-03  21.00.00    2,2       1376.0      80.0      9,2          948.0   
4 2004-10-03  22.00.00    1,6       1272.0      51.0      6,5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   

       AH  Unnamed: 15  Unnamed: 16  
0  0,7578          NaN          NaN  
1  0,7255          NaN          NaN  
2  0,7502          NaN          NaN  
3  0,7867          NaN          NaN  
4  0,7888          NaN          NaN

## Data Preprocessing
### **Dropping NaN values and columns**

In [4]:
data_set.dropna(how="all",axis=1,inplace=True)
data_set.dropna(how="all",axis=0,inplace=True)
data_set.shape

(9357, 15)

In [5]:
data_set.tail()

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  \
9352 2005-04-04  10.00.00    3,1       1314.0    -200.0     13,5   
9353 2005-04-04  11.00.00    2,4       1163.0    -200.0     11,4   
9354 2005-04-04  12.00.00    2,4       1142.0    -200.0     12,4   
9355 2005-04-04  13.00.00    2,1       1003.0    -200.0      9,5   
9356 2005-04-04  14.00.00    2,2       1071.0    -200.0     11,9   

      PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  \
9352         1101.0    472.0         539.0    190.0        1374.0   
9353         1027.0    353.0         604.0    179.0        1264.0   
9354         1063.0    293.0         603.0    175.0        1241.0   
9355          961.0    235.0         702.0    156.0        1041.0   
9356         1047.0    265.0         654.0    168.0        1129.0   

      PT08.S5(O3)     T    RH      AH  
9352       1729.0  21,9  29,3  0,7568  
9353       1269.0  24,3  23,7  0,7119  
9354       1092.0  26,9  18,3  0,6406  
9355        770.0  28,3  13,5  0,5139  
9356        816.0  28,5  13,1  0,5028

In [6]:
cols = list(data_set.columns[2:])
cols

['CO(GT)',
 'PT08.S1(CO)',
 'NMHC(GT)',
 'C6H6(GT)',
 'PT08.S2(NMHC)',
 'NOx(GT)',
 'PT08.S3(NOx)',
 'NO2(GT)',
 'PT08.S4(NO2)',
 'PT08.S5(O3)',
 'T',
 'RH',
 'AH']

Since decimal seperator by default is "," instead of "." we need to modify the float values.

In [7]:
for col in cols:
    if data_set[col].dtype != 'float64':
        str_x = pd.Series(data_set[col]).str.replace(',','.')
        float_X = []
        for value in str_x.values:
            fv = float(value)
            float_X.append(fv)

            data_set[col] = pd.DataFrame(float_X)

data_set.head()

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0 2004-10-03  18.00.00     2.6       1360.0     150.0      11.9   
1 2004-10-03  19.00.00     2.0       1292.0     112.0       9.4   
2 2004-10-03  20.00.00     2.2       1402.0      88.0       9.0   
3 2004-10-03  21.00.00     2.2       1376.0      80.0       9.2   
4 2004-10-03  22.00.00     1.6       1272.0      51.0       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1046.0    166.0        1056.0    113.0        1692.0       1268.0   
1          955.0    103.0        1174.0     92.0        1559.0        972.0   
2          939.0    131.0        1140.0    114.0        1555.0       1074.0   
3          948.0    172.0        1092.0    122.0        1584.0       1203.0   
4          836.0    131.0        1205.0    116.0        1490.0       1110.0   

      T    RH      AH  
0  13.6  48.9  0.7578  
1  13.3  47.7  0.7255  
2  11.9  54.0  0.7502  
3  11.0  60.0  0.7867  
4  11.2  59.6  0.7888

## Feature Selection

Here Date and Time are irrelevant to our prediction model

In [8]:
features=list(data_set.columns)

In [10]:
corr = data_set.corr()
corr.style.background_gradient(cmap='coolwarm')

In [9]:
features.remove('Date')
features.remove('Time')
features.remove('AH')
features

['CO(GT)',
 'PT08.S1(CO)',
 'NMHC(GT)',
 'C6H6(GT)',
 'PT08.S2(NMHC)',
 'NOx(GT)',
 'PT08.S3(NOx)',
 'NO2(GT)',
 'PT08.S4(NO2)',
 'PT08.S5(O3)',
 'T',
 'RH']

In [11]:
X = data_set[features]
Y = data_set['AH']

In [12]:
X.head()

CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  PT08.S2(NMHC)  NOx(GT)  \
0     2.6       1360.0     150.0      11.9         1046.0    166.0   
1     2.0       1292.0     112.0       9.4          955.0    103.0   
2     2.2       1402.0      88.0       9.0          939.0    131.0   
3     2.2       1376.0      80.0       9.2          948.0    172.0   
4     1.6       1272.0      51.0       6.5          836.0    131.0   

   PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  
0        1056.0    113.0        1692.0       1268.0  13.6  48.9  
1        1174.0     92.0        1559.0        972.0  13.3  47.7  
2        1140.0    114.0        1555.0       1074.0  11.9  54.0  
3        1092.0    122.0        1584.0       1203.0  11.0  60.0  
4        1205.0    116.0        1490.0       1110.0  11.2  59.6

### **Train Test Split**

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=0.4, random_state=0)

## Linear Regression Model

In [14]:
model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

### Error and Score

In [15]:
mse = mean_squared_error(y_test,y_pred)
rmse = mean_squared_error(y_test,y_pred,squared=False)
r2 = r2_score(y_test,y_pred)
print("MSE: ",mse)
print("RMSE: ",rmse)
print("r2: ",r2)

MSE:  0.9953986950559414
RMSE:  0.9976966949208268
r2:  0.999363823178606


**Comaprision of predicted results with actaul results**

In [17]:
df_preds = pd.DataFrame({'Actual': y_test.squeeze(), 'Predicted': y_pred.squeeze(), 'Loss': abs(y_test.squeeze()-y_pred.squeeze())})
df_preds

Actual  Predicted      Loss
1121  0.6821   0.044544  0.637556
8028  0.4129   1.176178  0.763278
3803  1.1558   1.718019  0.562219
4595  1.0899   1.914167  0.824267
587   1.0393   0.224666  0.814634
...      ...        ...       ...
4494  1.5116   1.638667  0.127067
4645  1.4039   1.065726  0.338174
2411  1.3500   1.476280  0.126280
7709  0.4127   1.046881  0.634181
4318  1.1406   1.052395  0.088205

[3743 rows x 3 columns]

## Polynomial Regression Model

In [18]:
polynomial_features= PolynomialFeatures(degree=2)
x_poly = polynomial_features.fit_transform(X_train)
model = LinearRegression()
model.fit(x_poly, y_train)

LinearRegression()

In [19]:
x_pred_poly = polynomial_features.fit_transform(X_test)
y_poly_pred = model.predict(x_pred_poly)

### Error and Score

In [24]:
pr_train,pr_test = model.score(x_poly , y_train), model.score(x_pred_poly , y_test)
print(f"Training Score: {pr_train}")
print(f"Test Score: {pr_test}")

Training Score: 0.9999994689091033
Test Score: 0.9999994199445779


In [20]:
mse = mean_squared_error(y_test,y_poly_pred)
rmse = mean_squared_error(y_test,y_poly_pred,squared=False)
r2 = r2_score(y_test,y_poly_pred)
print("MSE: ",mse)
print("RMSE: ",rmse)
print("r2: ",r2)

MSE:  0.000907587939192663
RMSE:  0.030126200211654026
r2:  0.9999994199445779


**Comaprision of predicted results with actaul results**

In [21]:
df_preds = pd.DataFrame({'Actual': y_test.squeeze(), 'Predicted': y_poly_pred.squeeze(), 'Loss': abs(y_test.squeeze()-y_poly_pred.squeeze())})
df_preds

Actual  Predicted      Loss
1121  0.6821   0.689703  0.007603
8028  0.4129   0.399739  0.013161
3803  1.1558   1.153307  0.002493
4595  1.0899   1.068687  0.021213
587   1.0393   1.057095  0.017795
...      ...        ...       ...
4494  1.5116   1.475953  0.035647
4645  1.4039   1.423661  0.019761
2411  1.3500   1.372531  0.022531
7709  0.4127   0.427434  0.014734
4318  1.1406   1.168856  0.028256

[3743 rows x 3 columns]

## Conclusion

In this experiment we have modeled polynomial regression to predict absolute humidity in atmosphere.

The resulting scores from polynomial regression are:
- **MSE**:  0.000907587939192663
- **RMSE**:  0.030126200211654026
- **r2**:  0.9999994199445779

## Comparing with Linear regression

- **MSE**:  0.9953986950559414
- **RMSE**:  0.9976966949208268
- **r2**:  0.999363823178606
